### ISSUE TITLE
Normal.icdf is differs on different cuda devices

### ISSUE THREAD

**Issue by sdaulton** [pytorch/pytorch#38889]
## 🐛 Bug

Normal.icdf is different on different cuda devices

[normal_cdf_bug.ipynb.txt](https://github.com/pytorch/pytorch/files/4664656/normal_cdf_bug.ipynb.txt)

```
import torch
x = torch.rand(10, dtype=torch.double)
normal = torch.distributions.Normal(0,1)
normal.icdf(x)
```
Output: 
>>> tensor([-0.0274,  0.9183,  0.8124,  0.1397, -0.1859,  0.0392, -0.3418,  0.8999,
         2.2092, -0.5433], dtype=torch.float64)

```
normal = torch.distributions.Normal(0,1)
normal.icdf(x.cuda())
```
Output:
>>> tensor([-0.0274,  0.9183,  0.8124,  0.1397, -0.1859,  0.0392, -0.3418,  0.8999,
     2.2092, -0.5433], dtype=torch.float64)
```
normal = torch.distributions.Normal(0,1)
normal.icdf(x.cuda(1))
```
Output:
>>> tensor([-0.0219,  0.6415,  0.5835,  0.1111, -0.1475,  0.0313, -0.2675,  0.6318,
         0.9728, -0.4131], device='cuda:1', dtype=torch.float64)
## Expected behavior

I expected Normal.icdf to be the same on all devices.

## Environment
Collecting environment information...
PyTorch version: 1.6.0a0-fb
Is debug build: No
CUDA used to build PyTorch: 9.2.88

OS: CentOS Linux 7 (Core)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-39)
CMake version: Could not collect

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: Tesla V100-SXM2-16GB
GPU 1: Tesla V100-SXM2-16GB
GPU 2: Tesla V100-SXM2-16GB
GPU 3: Tesla V100-SXM2-16GB
GPU 4: Tesla V100-SXM2-16GB
GPU 5: Tesla V100-SXM2-16GB
GPU 6: Tesla V100-SXM2-16GB
GPU 7: Tesla V100-SXM2-16GB

Nvidia driver version: 396.69
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] Could not collect
[conda] Could not collect

cc @ezyang @gchanan @zou3519 @ngimel

**Comment by ShawnZhong** [comment]
I don’t have multiple GPUs on hand to verify this, but my guess is to call https://pytorch.org/docs/master/cuda.html#torch.cuda.manual_seed_all first to set the same random seed across all GPUs.

**Comment by SsnL** [comment]
@ShawnZhong this is a deterministic operation, so random seed should matter.

**Comment by sdaulton** [comment]
shouldn't*

**Comment by thoglu** [comment]
the first device agrees with cpu right?

**Comment by ngimel** [comment]
There's something wrong with how we are setting device guard. Easy repro without distributions:
```
import torch

x = torch.rand(10, dtype=torch.double, device="cuda:1")
val = torch.tensor(1.)
y=x * val
torch.cuda.synchronize()
print("ok")
y=val * x
torch.cuda.synchronize()
print("ok") #not ok, illegal memory access
```

**Comment by ngimel** [comment]
@sdaulton if you need a quick workaround, you can do 
```
import torch
x = torch.rand(10, dtype=torch.double, device="cuda:1")
normal = torch.distributions.Normal(0,1)
with torch.cuda.device_of(x):
    normal.icdf(x)
```

**Comment by kurtamohler** [comment]
I found some more details about this issue, but I don't yet know what the solution is. The functions in `build/aten/src/ATen/CUDAType.cpp` create an OptionalDeviceGuard to match the device of `self`. For instance:
```
 Tensor mul_Tensor(const Tensor & self, const Tensor & other) { 
     const OptionalDeviceGuard device_guard(device_of(self));
     return at::native::mul(self, other);
 }
```

In @ngimel's repro above, for the `val * x` case, `self` will be the scalar `val`. Scalar tensors have the device type `cpu`, so the OptionalDeviceGuard will be for cpu, not cuda. This problem doesn't cause an issue (seemingly accidentally?) in the usual case where `x.device()` is `cuda:0`, because PyTorch makes `cudaGetDevice()` default to index 0. But for the same reason, the `cuda:1` case breaks because the OptionalDeviceGuard does not know that we should be switching over to `cuda:1`, so when the `mul` operation is executed, we end up trying to access a `cuda:1` buffer from `cuda:0`. As @ngimel showed, we can work around the issue by explicitly telling pytorch to use `cuda:1`. But if we tried to do something like `torch.tensor(1.) * torch.tensor(..., device="cuda:0")` inside of the `with torch.cuda.device_of(x)` in her code snippet, it would break for the same reason.

Here are some examples that might help explain what I mean:

This `cuda:1` operation **fails** because `x` doesn't match `torch.cuda.set_device()` setting:
```
torch.cuda.set_device(0)
x = torch.rand(10, dtype=torch.double, device="cuda:1")
val = torch.tensor(1.)
y = val * x
torch.cuda.synchronize()
```

This `cuda:0` operation **fails** because `x` doesn't match `torch.cuda.set_device()` setting:
```
torch.cuda.set_device(1)
x = torch.rand(10, dtype=torch.double, device="cuda:0")
val = torch.tensor(1.)
y = val * x
torch.cuda.synchronize()
```

This `cuda:0` operation **works** because `x` does match `torch.cuda.set_device()` setting:
```
torch.cuda.set_device(0)
x = torch.rand(10, dtype=torch.double, device="cuda:0")
val = torch.tensor(1.)
y = val * x
torch.cuda.synchronize()
```

This `cuda:1` operation **works** because `x` does match `torch.cuda.set_device()` setting:
```
torch.cuda.set_device(1)
x = torch.rand(10, dtype=torch.double, device="cuda:1")
val = torch.tensor(1.)
y = val * x
torch.cuda.synchronize()
```

It seems like we need something either before or inside of `mul_Tensor()` (and all other functions in `CUDAType.cpp`) that recognizes when `self` is a scalar, and then create a DeviceGuard from `other` instead. I'm not sure how that should be done in a generalized way that applies to all the operations. I might need some help with thinking of a solution.

**Comment by ngimel** [comment]
Since this situation only occurs when gpu_kernel_with_scalars is used, a general solution that changes the codegen to decide where to put device guard seems like an overkill, and probably we can put an additional deviceGuard here https://github.com/pytorch/pytorch/blob/d627f2b174136b720dcc99380e2b98f535d1f169/aten/src/ATen/native/cuda/Loops.cuh#L79-L107.

**Comment by kurtamohler** [comment]
Thanks @ngimel ! I made a patch.
